src/wp-includes/formatting.php

```diff
@@ -3351,7 +3351,7 @@ function wp_remove_targeted_link_rel_filters() {
 function translate_smiley( $matches ) {
 	global $wpsmiliestrans;
 
-	if ( count( $matches ) == 0 ) {
+	if ( ! $matches || count( $matches ) == 0 ) {
```

Comment:
```suggestion
	if ( ! is_array( $matches ) || count( $matches ) == 0 ) {
```
 or maybe (`is_countable()` is polyfilled):
 ```suggestion
	if ( ! is_countable( $matches ) || count( $matches ) == 0 ) {
```

... otherwise the function would still try to act on a non-empty string, `true` etc., which _could_/would still lead to errors/notices.

Also note that `reset()` is used on the next line, so the check may need to be extended further to also check for `$matches` being iterable.

In other words: before a fix could be considered, tests needs to be added with different parameter types being passed as `$matches`. These tests needs to include a class implementing `Countable`, a class implementing `Traversable` and a class implementing both.
Based on whether or not those are handled without errors on PHP < 8.0 without any fixes, the most appropriate fix should be determined.

Reply:
Nicely spotted.

Since `$matches` is documented as an array, I think we can go ahead and use `! is_array( $matches )`. This would also preserve the documented behaviour of returning a string (rather than type hinting `array`, for example).

It's _possible_ that some have passed a countable/traversable object instead of an array to `translate_smiley()`, which would make array enforcement a BC break. However, as hesitant as I am to break BC, I don't believe we have any tickets or examples from the wild that suggest a need for us to facilitate that usage.

As an aside, I think polyfilling `is_countable()` is a good move if we aren't likely to bump the minimum PHP version to 7.3+ in the near future, though that would be a separate ticket to make sure it has its own history documented.

Reply:
> Since `$matches` is documented as an array, I think we can go ahead and use `! is_array( $matches )`.

To be sure, it would probably be a good idea to do a cursory review of how this function is used in the "wild" (i.e. do a search of the plugin/theme directory and review what comes back).

> This would also preserve the documented behaviour of returning a string (rather than type hinting `array`, for example).

Type hinting the parameter as `array` would not change the return type, but would make the function throw catchable exceptions when passed something which isn't an array, which would be a bigger BC-break.

Similarly, casting to `array` would not change the return type, but would allow the function to act on input it shouldn't be acting on in the first place, so would not be the correct way to handle this.

> As an aside, I think polyfilling `is_countable()` is a good move if we aren't likely to bump the minimum PHP version to 7.3+ in the near future, though that would be a separate ticket to make sure it has its own history documented.

It already **_is_** polyfilled: https://core.trac.wordpress.org/browser/trunk/src/wp-includes/compat.php#L352
Same goes for `is_iterable()`: https://core.trac.wordpress.org/browser/trunk/src/wp-includes/compat.php#L373

Reply:
> To be sure, it would probably be a good idea to do a cursory review of how this function is used in the "wild" (i.e. do a search of the plugin/theme directory and review what comes back).

Agreed.

Plugins: [16 matches, none appear to pass non-arrays](https://wpdirectory.net/search/01FSMDD1AYQYCSZYST86T62TZJ)
Themes: [0 matches](https://wpdirectory.net/search/01FSMDNFJWQE356DVPGY0F13D2)

> Type hinting the parameter as array would not change the return type, but would make the function throw catchable exceptions when passed something which isn't an array, which would be a bigger BC-break.

Yep, that's what I was referring to: causing it to throw exceptions would change the behaviour.

> Similarly, casting to array would not change the return type, but would allow the function to act on input it shouldn't be acting on in the first place, so would not be the correct way to handle this.

Hmm... now that you've brought it up, a few thoughts that hadn't occurred to me:
- The function is called `translate_smiley`, not `translate_smilies`.
- The docblock states:
> Looks up one smiley code in the $wpsmiliestrans global array and returns an `<img>` string for that smiley.

To me, this suggests that `translate_smiley()` should have always accepted `array` or `string`. Comparing the function's description with its code, it seems that the function was named and documented for a general case, but coded for a specific case, i.e. `Callback handler for convert_smilies()`. Though, this sounds more like a separate `enhancement` ticket than resolving the reported issue.

Back to validating:
The function assumes that the smiley key already exists. Nothing is checked, it's actually just blindly retrieved. The function therefore requires that the smiley has already been looked up in `$wpsmiliestrans` prior to being called.

Given that there are at least 16 plugins that reference `translate_smiley` (although a number of these are copy-paste for custom implementations), do you think it's worth considering an extra guard to check that `wpsmiliestrans` actually has the key, then return early, possibly with an empty string? Or allow PHP native to error out on the missing key?

> or maybe (is_countable() is polyfilled):
> It already is polyfilled:

_notices parentheses_
Woops! My bad!

Reply:
> Hmm... now that you've brought it up, a few thoughts that hadn't occurred to me:
> 
>     * The function is called `translate_smiley`, not `translate_smilies`.
> 
>     * The docblock states:
> 
> 
> > Looks up one smiley code in the $wpsmiliestrans global array and returns an `<img>` string for that smiley.
> 
> To me, this suggests that `translate_smiley()` should have always accepted `array` (and `reset()` as it currently does) _or_ `string`. Comparing the function's description with its code, it seems that the function was named and documented for a general case, but coded for a specific case, i.e. `Callback handler for convert_smilies()`. Though, this sounds more like a separate `enhancement` ticket than resolving the reported issue.

No. The way I read it, the function is intended as a callback for `preg_replace_callback()` and expects to be passed an array which represents _one_ "match" as taken from an array of matches.
So _string_ was never a valid input, it always expected an array.

You can confirm this for yourself by looking at the code for [convert_smilies()](https://developer.wordpress.org/reference/functions/convert_smilies/).

> Back to validating: The function assumes that the smiley key already exists. Nothing is checked, it's actually just blindly retrieved. The function therefore requires that the smiley has already been looked up in `$wpsmiliestrans` prior to being called.
> 
> Given that there are at least 16 plugins that reference `translate_smiley` (although a number of these are copy-paste for custom implementations), do you think it's worth considering an extra guard to check that `$wpsmiliestrans` actually has the key, then return early, possibly with an empty string? Or allow PHP native to error out on the missing key?

Not really. As per the above, the function is supposed to be used only as a callback in a specific situation where the regex in `convert_smilies()` has already confirmed that the key will exist (otherwise, there would be no match).

If plugins are _doing it wrong_, they should get a PHP notice and fix their own code.

On a more architectural note: this should never have been a global/public function. This should have been a `private` method in a class. Unfortunately, that's not something we can change anymore now.

Reply:
To summarise a discussion in Slack DM:

While this function is intended only as a callback for `convert_smilies()`, it's a public function that can be, and is, used directly by extenders. Therefore, to be on the safe side, we agreed that implementing `! is_array( $matches )` was appropriate.